invertase/react-native-firebase
# feat: Added Expo config plugin

### Description

Expo released [development clients](https://docs.expo.io/clients/introduction/) and rolled out an interface called [config plugins](https://docs.expo.io/guides/config-plugins/). These allow users to add native modules which aren't included in Expo Go.

This PR adds a config plugin, which allows users for painless setup of RNFB, without touching any native code. Until now, I've been maintaining the plugin as a separate npm package: [`with-rn-firebase`](https://www.npmjs.com/package/with-rn-firebase) and had some positive feedback. Now it's time to integrate it into the upstream library 😄  For more context, see the discussion at #5386. 

Basically I've copied my code from https://github.com/barthap/with-rn-firebase, but I've split it into separate plugins for each affected package to avoid additional config like setting `enableCrashlytics` etc.

**RNFB repo related changes:**
- Added `build:plugin` and `lint:plugin` npm scripts for affected packages
- The `build:plugin` is run during `npm prepare` phase
- Added `plugin/__tests__` directories with some basic tests, they work when running `npm run tests:jest`
- I had to modify global `tsconfig.json` to add `esModuleInterop: true` - otherwise the plugin tests failed. I haven't seen any regressions and hope it's not breaking - please correct me if it is.
- Added dependency `@expo/config-plugins` and devDependency with plugin tsconfig base.
- Added `packages/**/plugin/build` dir to eslint and prettier ignore - it's a generated directory
- Added some values to `sdkVersions` section in `app/package.json`

For the plugin to work, the package should have a `app.plugin.js` file in its root folder - in this case it only delegates to the `plugin/build` directory. The `build` directory is gitignored and generated during `npm build:plugin` from TypeScript files inside `plugin/src`. 

#### The plugins

As for now, to reflect current `with-rn-firebase` functionality, I've added plugins for 3 packages:
- **App**:
   Its purpose is to copy `google-services.json` and `GoogleServices-Info.plist` into correct places, modify `AppDelegate.m`, and add Gradle dependency and plugin - basically the installation steps from the homepage.
  
   The Google Services files by default are copied from the paths specified by `expo.ios|android.googleServicesFile` in `app.json`.
- **Crashlytics**:
   Installs the required Gradle dependencies for Android. It doesn't yet support the optional release NDK installation step.
- **Performance Monitoring**:
   Installs the required Gradle dependency for Android.


#### Usage / How it works

When using Expo, replace the `yarn add`/`npm install` with `expo install`:
```
expo install @react-native-firebase/app
expo install @react-native-firebase/crashlytics
```
Then in the `app.json` the `plugins` section should be updated automatically. When not using `expo install`, you have to add the plugins on your own`.

Remember to configure the `expo.android.googleServicesFile` and `expo.ios.googleServicesFile` in `app.json`. The paths are relative to the project root.

The `app.json` file may look like this (non-related parts omitted):
```json
{
  "expo": {
    "android": {
      "googleServicesFile": "./google-services.json"
    },
    "ios": {
       "googleServicesFile": "./GoogleService-Info.plist"
    },
    "plugins": [
      "@react-native-firebase/app",
      "@react-native-firebase/crashlytics"
    ]
  }
}
```

#### Further steps

After this PR is merged:

- Update the [docs `Expo` section](https://rnfirebase.io/#expo) and possibly the installation steps for the packages: `app`, `perf`, `crashlytics`

These were not included in my original plugin, they're not implemented yet:
- Investigate the need of config plugins for the `FCM` module (see the discussion in https://github.com/barthap/with-rn-firebase/issues/6)
- Add option to install NDK Crashlytics in release build
- Investigate the need of adding plugin for other packages. Afaik there are some configuration options requiring the `Podfile` changes, which are not trivial for config plugins.

### Related issues

Discussion #5386 

### Release Summary

- **app, crashlytics, perf**: added Expo config plugins

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- My change supports the following platforms;
- My change includes tests;
  - [x] `jest` tests added or updated in `packages/\*\*/__tests__` (_added tests for config plugins_)
- This is a breaking change;


### Test Plan

Steps that anyone should be able to reproduce:

1. Installed npm RNFB packages to a separate directory and applied my plugin changes - I've attached the packages to this post:
   [patched-packages.zip](https://github.com/invertase/react-native-firebase/files/6782106/patched-packages.zip)
2. In another directory `expo init app_name_here` - selected managed template
   > You can use `npx crna -t with-dev-client` instead to try it out with [Expo development client](https://docs.expo.io/clients/introduction/)
3. Run `expo prebuild` on this project
4. `git commit` your changes
5. `yarn link` all 3 RNFB packages into Expo app project
4. Add `plugins` section to `app.json`. If you have [Expo Tools](https://marketplace.visualstudio.com/items?itemName=byCedric.vscode-expo) VSCode extension, the intellisense should propose them automatically 
   ![Screenshot 2021-07-08 at 08 32 56](https://user-images.githubusercontent.com/278340/124873736-25c3b180-dfc7-11eb-8ab4-a898696d59af.png)
5. Copy or create blank `google-services.json` and `GoogleService-Info.plist` (e.g. in app root dir) and specify the `expo.android.googleServicesFile` and `expo.ios.googleServicesFile` fields in `app.json`.
6. `git commit` again to have a checkpoint to easy reset
7. Run `expo prebuild` and observe the git diff. Experiment with different combinations of `plugins` in app.json
   > You can build and run your app with `expo run:ios` / `expo run:android`
